### PR TITLE
Add arange op

### PR DESCRIPTION
**Pull request by WashimNeupane**
## Description
Added torch.arange support for pytorch frontend. 

1. Added ArangeParser in creation.py to handle aten.arange.default, aten.arange.start and aten.arange.start_step
2. Expose add_arange method in python bindings
3. Introduces new ArangeNode in sdfg ir. 
4. Implemnets graph exansion for the ArangeNode
5. Tests and validators for verification

## Related Issues

- Closes #

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal change (no functional change)
- [ ] Documentation update

## Quality Checklist
*Please ensure the following are completed before requesting review:*

* [x] Runtime-compatible (no externally visible existing method has its parameters changed, was deleted or has its exception behavior changed)
* [x] Compile-time compatible (new arguments & properties added have default values, but users need to recompile against the new version to work)

IR:
* [ ] Introduces new types into the SDFG IR
* [x]  Changed / added properties on SDFG IR Elements (LibraryNodes, ControlFlowNodes DataFlowNodes etc.)
  *  [x] New / modified types have serialization / deserialization support
  *  [x] Deserialization is backward compatible (new serialization can deserialize old format and represent it in the new format if applicable)
  *  [x] Serialized format is backward compatible (previous deserializer can read new serialized format without incorrectness (just lacking additional guarantees that are optional or did not exist before)
  *  [ ] Serializer has an option to emit the previous format

### Unit Tests
- [x] New unit tests have been added for the changes.
- [x] Existing unit tests pass locally.
- [x] Edge cases and error paths are covered.

### Integration Tests
- [x] New integration tests have been added (or existing ones updated).
- [x] Integration tests pass locally.
- [x] Interactions with other components/services are verified.

### Documentation
- [ ] Public APIs, options, and behavior changes are documented.
- [ ] README / docs site / inline comments updated where applicable.
- [ ] Changelog / release notes updated (if applicable).

### Test Coverage
- [x] Test coverage has not decreased as a result of this change.
- [x] Coverage report reviewed and acceptable.
- [x] Critical / new code paths are covered.

## How Has This Been Tested?
- Ran pytest -v -s pytorch/tests/ops/test_creation.py locally.
- Verified test passes for test_arange_default, test_arange_start, and test_arange_start_step.
- Dumped the underlying generated DOT graphs and JSON intermediate representation to visually verify that the _arange_tmp intra-iteration data dependency correctly maps out a fully connected non-cyclical DAG.

## Additional Context
Add any other context about the PR here (design decisions, trade-offs, follow-ups).

### CODE REVIEW

File: pytorch/docc/pytorch/graph_parser/creation.py [hunk 1]
```diff
@@ -129,3 +129,87 @@ def parse(
...
+            step_container,
+            step_type,
+            result_container,
+            result_tensor,
+            debug_info,
+        )
+
+
+register_module("aten.arange.default", ArangeParser())
+register_module("aten.arange.start", ArangeParser())
```
Comment (Atrisan) [inline pytorch/docc/pytorch/graph_parser/creation.py]: aten.arange(start/default) do not exist

File: pytorch/docc/pytorch/graph_parser/creation.py [hunk 2]
```diff
@@ -129,3 +129,87 @@ def parse(
...
+                raise GraphParserError(
+                    self,
+                    node,
+                    "Expected 2 or 3 arguments but got " + str(len(node.args)),
+                )
+            start_val = self.get_arg_sdfg_value(node, container_info, 0)
+            step_val = (
+                self.get_arg_sdfg_value(node, container_info, 2)
+                if len(node.args) == 3
+                else ("1", Scalar(PrimitiveType.Int64))
```
Comment (Atrisan) [inline pytorch/docc/pytorch/graph_parser/creation.py]: arange.start_step always gets 3 arguments + kwargs:
arange.start_step(Scalar start, Scalar end, Scalar step=1, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor

File: python/bindings/builder/py_structured_sdfg_builder.h [hunk 3]
```diff
@@ -419,6 +419,16 @@ class PyStructuredSDFGBuilder {
...
 
+    void add_arange(
+        const std::string& start,
+        const sdfg::types::Scalar& start_type,
+        const std::string& step,
+        const sdfg::types::Scalar& step_type,
+        const std::string& out,
+        const sdfg::types::Tensor& out_type,
+        const sdfg::DebugInfo& debug_info = sdfg::DebugInfo()
+    );
```
Comment (Atrisan) [inline python/bindings/builder/py_structured_sdfg_builder.h:432]: the end argument is missing for a correct translation

File: python/bindings/bindings.cpp [hunk 4]
```diff
@@ -758,6 +758,17 @@ PYBIND11_MODULE(_sdfg, m) {
...
+        .def(
+            "add_arange",
+            &PyStructuredSDFGBuilder::add_arange,
+            py::arg("start"),
+            py::arg("start_type"),
+            py::arg("step"),
+            py::arg("step_type"),
+            py::arg("out"),
+            py::arg("out_type"),
+            py::arg("debug_info") = sdfg::DebugInfo()
```
Comment (Atrisan) [inline python/bindings/bindings.cpp:772]: the end argument is missing for a correct translation

File: python/bindings/builder/py_structured_sdfg_builder.cpp [hunk 5]
```diff
@@ -1832,6 +1833,38 @@ void PyStructuredSDFGBuilder::add_fill_op(
...
 
+void PyStructuredSDFGBuilder::add_arange(
+    const std::string& start,
+    const sdfg::types::Scalar& start_type,
+    const std::string& step,
+    const sdfg::types::Scalar& step_type,
+    const std::string& out,
+    const sdfg::types::Tensor& out_type,
+    const sdfg::DebugInfo& debug_info
+) {
```
Comment (Atrisan) [inline python/bindings/builder/py_structured_sdfg_builder.cpp:1846]: the end argument is missing for a correct translation

File: pytorch/tests/ops/test_creation.py [hunk 6]
```diff
@@ -49,3 +49,54 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
...
+
+    check(ArangeFloatNet(), *(), target=target)
+
+
+def test_arange_default_dtype(target: str) -> None:
+    class ArangeDefaultDtypeNet(nn.Module):
+        def forward(self) -> torch.Tensor:
+            return torch.arange(8, dtype=torch.float32)
+
+    check(ArangeDefaultDtypeNet(), *(), target=target)
```
Comment (Atrisan) [inline pytorch/tests/ops/test_creation.py:102]: what about test for the remaining kwargs:
layout, device, pin_memory?

File: pytorch/docc/pytorch/graph_parser/creation.py [hunk 7]
```diff
@@ -129,3 +129,87 @@ def parse(
...
+class ArangeParser(GraphParserModule):
+    def parse(
+        self,
+        node: torch.fx.Node,
+        builder: StructuredSDFGBuilder,
+        container_info: ContainerInfos,
+    ) -> None:
+        if not set(node.kwargs.keys()).issubset(
+            {"dtype", "layout", "device", "pin_memory", "memory_format"}
+        ):
```
Comment (Atrisan) [inline pytorch/docc/pytorch/graph_parser/creation.py:151]: memory format is not supported

File: pytorch/docc/pytorch/graph_parser/creation.py [hunk 8]
```diff
@@ -129,3 +129,87 @@ def parse(
...
+            start_val = self.get_arg_sdfg_value(node, container_info, 0)
+            step_val = (
+                self.get_arg_sdfg_value(node, container_info, 2)
+                if len(node.args) == 3
+                else ("1", Scalar(PrimitiveType.Int64))
+            )
+        else:
+            raise GraphParserError(self, node, f"Unsupported target {node.target}")
+
+        start_container = start_val if isinstance(start_val, str) else start_val[0]
```
Comment (Atrisan) [inline pytorch/docc/pytorch/graph_parser/creation.py:229]: kwarg checks are missing

File: sdfg/include/sdfg/data_flow/library_nodes/math/tensor/arange_node.h [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        data_flow::DataFlowGraph& parent,
+        const std::vector<symbolic::Expression>& shape,
+        const data_flow::ImplementationType& impl_type = data_flow::ImplementationType_NONE
+    );
+
+    // In-edge indices: _out is the result buffer pointer (written to), _start and _step are read-only scalar inputs.
+    // All three are modelled as in-edges following the ElementWiseDataflowTensorNode convention.
+    static auto constexpr RESULT_PTR_IDX = 0; // connector: _out
+    static auto constexpr START_IDX = 1; // connector: _start
+    static auto constexpr STEP_IDX = 2; // connector: _step
```
Comment (Atrisan) [inline sdfg/include/sdfg/data_flow/library_nodes/math/tensor/arange_node.h]: end missing

File: sdfg/src/data_flow/library_nodes/math/tensor/arange_node.cpp [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+        }
+    }
+    if (edges.size() > STEP_IDX && edges[STEP_IDX] != nullptr) {
+        auto* step_edge = edges.at(STEP_IDX);
+        if (step_edge->base_type().type_id() != types::TypeID::Scalar) {
+            throw InvalidSDFGException(
+                "ArangeNode: _step input must be of scalar type. Found type: " + step_edge->base_type().print()
+            );
+        }
+    }
```
Comment (Atrisan) [inline sdfg/src/data_flow/library_nodes/math/tensor/arange_node.cpp:75]: check end

File: sdfg/src/data_flow/library_nodes/math/tensor/arange_node.cpp [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+    for (auto& dim : shape_) {
+        dim = symbolic::subs(dim, replacements);
+    }
+}
+
+bool ArangeNode::supports_integer_types() const { return true; }
+
+passes::LibNodeExpander::ExpandOutcome ArangeNode::
+    expand(passes::LibNodeExpander::ExpandContext& context, structured_control_flow::Block& block) {
+    auto& dataflow = this->get_parent();
```
Comment (Atrisan) [inline sdfg/src/data_flow/library_nodes/math/tensor/arange_node.cpp:104]: Add support for end to have consistent behaviour

**Review by Atrisan — CHANGES_REQUESTED** [review]
Check against the op definition again to correctly support all parameters, especially end.

aten.arange.start_step
	

arange.start_step(Scalar start, Scalar end, Scalar step=1, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor

**Review by Atrisan — APPROVED** [review]
